File: macos/Sources/MacDroidSyncCore/Protocol.swift
import Foundation

/// Wire protocol shared by the macOS and the Android side.
///
/// Frame layout:  [4 bytes big-endian length N][1 byte kind][N - 1 bytes body]
/// The only frame allowed in plaintext is the server challenge; every other
/// frame carries an AES-256-GCM sealed box (see `FrameCodec`).
public enum Wire {
    public static let version = 1
    public static let defaultPort: UInt16 = 47831
    public static let bonjourServiceType = "_macdroidsync._tcp"

    /// Hard limit for a single frame, protects against a hostile length prefix.
    public static let maxFrameSize = 4 * 1024 * 1024
    /// Clipboard payloads above this size are skipped instead of being synced.
    public static let maxClipboardBytes = 512 * 1024

    public static let heartbeatInterval: TimeInterval = 15
    public static let receiveTimeout: TimeInterval = 30
}

public enum FrameKind: UInt8 {
    case plaintext = 0x01
    case encrypted = 0x02
}

public enum MessageType {
    public static let challenge = "challenge"
    public static let hello = "hello"
    public static let helloAck = "hello-ack"
    public static let clipboard = "clipboard"
    public static let clipboardAck = "clipboard-ack"
    public static let requestClipboard = "request-clipboard"
    public static let ping = "ping"
    public static let pong = "pong"
    public static let heartbeat = "heartbeat"
    public static let bye = "bye"
}

/// One protocol message. Absent fields are omitted from the JSON payload.
public struct Message: Codable {
    public var v: Int
    public var seq: UInt64
    public var type: String
    public var ts: Int64
    public var text: String?
    public var device: String?
    public var deviceId: String?
    public var challenge: String?
    public var token: UInt64?
    public var reason: String?

    public init(
        v: Int = Wire.version,
        seq: UInt64 = 0,
        type: String,
        ts: Int64 = Message.now(),
        text: String? = nil,
        device: String? = nil,
        deviceId: String? = nil,
        challenge: String? = nil,
        token: UInt64? = nil,
        reason: String? = nil
    ) {
        self.v = v
        self.seq = seq
        self.type = type
        self.ts = ts
        self.text = text
        self.device = device
        self.deviceId = deviceId
        self.challenge = challenge
        self.token = token
        self.reason = reason
    }

    public static func now() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }

    public func encoded() throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return try encoder.encode(self)
    }

    public static func decode(_ data: Data) throws -> Message {
        try JSONDecoder().decode(Message.self, from: data)
    }
}

public enum ProtocolError: LocalizedError {
    case frameTooLarge(Int)
    case malformedFrame(String)
    case unexpectedFrame(String)
    case authenticationFailed
    case replayDetected(UInt64)

    public var errorDescription: String? {
        switch self {
        case .frameTooLarge(let size):
            return "Frame of \(size) bytes exceeds the maximum frame size"
        case .malformedFrame(let detail):
            return "Malformed frame: \(detail)"
        case .unexpectedFrame(let detail):
            return "Unexpected frame: \(detail)"
        case .authenticationFailed:
            return "Authentication failed - the pairing code does not match"
        case .replayDetected(let seq):
            return "Replayed or out of order message (seq \(seq))"
        }
    }
}

/// Builds and parses the length-prefixed framing used on the socket.
public enum Framing {
    public static func frame(kind: FrameKind, body: Data) -> Data {
        var out = Data(capacity: body.count + 5)
        let length = UInt32(body.count + 1)
        out.append(UInt8((length >> 24) & 0xFF))
        out.append(UInt8((length >> 16) & 0xFF))
        out.append(UInt8((length >> 8) & 0xFF))
        out.append(UInt8(length & 0xFF))
        out.append(kind.rawValue)
        out.append(body)
        return out
    }

    /// Pulls the first complete frame out of `buffer`, removing its bytes.
    /// Returns nil when more bytes are needed.
    public static func nextFrame(from buffer: inout Data) throws -> (kind: FrameKind, body: Data)? {
        guard buffer.count >= 4 else { return nil }
        let header = [UInt8](buffer.prefix(4))
        let length = (Int(header[0]) << 24) | (Int(header[1]) << 16) | (Int(header[2]) << 8) | Int(header[3])
        guard length >= 1 else { throw ProtocolError.malformedFrame("zero length") }
        guard length <= Wire.maxFrameSize else { throw ProtocolError.frameTooLarge(length) }
        guard buffer.count >= 4 + length else { return nil }

        let payload = Data(buffer[buffer.startIndex + 4 ..< buffer.startIndex + 4 + length])
        buffer.removeFirst(4 + length)

        guard let kind = FrameKind(rawValue: payload[payload.startIndex]) else {
            throw ProtocolError.malformedFrame("unknown frame kind \(payload[payload.startIndex])")
        }
        return (kind, Data(payload.dropFirst()))
    }
}

File: android/app/src/main/java/pl/wojas/macdroidsync/Protocol.kt
package pl.wojas.macdroidsync

import org.json.JSONObject

/** Wire constants shared with the macOS app, see PROTOCOL.md. */
object Wire {
    const val VERSION = 1
    const val DEFAULT_PORT = 47831
    const val SERVICE_TYPE = "_macdroidsync._tcp"

    const val MAX_FRAME_SIZE = 4 * 1024 * 1024
    const val MAX_CLIPBOARD_BYTES = 512 * 1024

    /** Raw bytes per file-chunk; base64 grows it by a third. */
    const val FILE_CHUNK_BYTES = 192 * 1024
    /** The Mac refuses anything bigger, so there is no point in starting. */
    const val MAX_FILE_BYTES = 512L * 1024 * 1024

    const val HEARTBEAT_INTERVAL_MS = 15_000L
    const val RECEIVE_TIMEOUT_MS = 30_000L

    const val KIND_PLAINTEXT: Byte = 0x01
    const val KIND_ENCRYPTED: Byte = 0x02
}

object MessageType {
    const val CHALLENGE = "challenge"
    const val HELLO = "hello"
    const val HELLO_ACK = "hello-ack"
    const val CLIPBOARD = "clipboard"
    const val CLIPBOARD_ACK = "clipboard-ack"
    const val REQUEST_CLIPBOARD = "request-clipboard"
    const val PING = "ping"
    const val PONG = "pong"
    const val HEARTBEAT = "heartbeat"
    const val BYE = "bye"
    const val FILE_OFFER = "file-offer"
    const val FILE_CHUNK = "file-chunk"
    const val FILE_END = "file-end"
    const val FILE_ACK = "file-ack"
}

/** One protocol message; absent fields are left out of the JSON payload. */
data class Message(
    val type: String,
    val seq: Long = 0,
    val v: Int = Wire.VERSION,
    val ts: Long = System.currentTimeMillis(),
    val text: String? = null,
    val device: String? = null,
    val deviceId: String? = null,
    val challenge: String? = null,
    val token: Long? = null,
    val reason: String? = null,
    // File transfer, see PROTOCOL.md section 6.
    val fileId: String? = null,
    val name: String? = null,
    val size: Long? = null,
    val mime: String? = null,
    val sha256: String? = null,
    /** Base64 of one file-chunk payload. */
    val data: String? = null,
    val ok: Boolean? = null,
    val path: String? = null,
) {
    fun toBytes(): ByteArray {
        val json = JSONObject()
        json.put("v", v)
        json.put("seq", seq)
        json.put("type", type)
        json.put("ts", ts)
        text?.let { json.put("text", it) }
        device?.let { json.put("device", it) }
        deviceId?.let { json.put("deviceId", it) }
        challenge?.let { json.put("challenge", it) }
        token?.let { json.put("token", it) }
        reason?.let { json.put("reason", it) }
        fileId?.let { json.put("fileId", it) }
        name?.let { json.put("name", it) }
        size?.let { json.put("size", it) }
        mime?.let { json.put("mime", it) }
        sha256?.let { json.put("sha256", it) }
        data?.let { json.put("data", it) }
        ok?.let { json.put("ok", it) }
        path?.let { json.put("path", it) }
        return json.toString().toByteArray(Charsets.UTF_8)
    }

    companion object {
        fun parse(bytes: ByteArray): Message {
            val json = JSONObject(String(bytes, Charsets.UTF_8))
            return Message(
                type = json.optString("type"),
                seq = json.optLong("seq"),
                v = json.optInt("v", Wire.VERSION),
                ts = json.optLong("ts"),
                text = json.optStringOrNull("text"),
                device = json.optStringOrNull("device"),
                deviceId = json.optStringOrNull("deviceId"),
                challenge = json.optStringOrNull("challenge"),
                token = if (json.has("token")) json.optLong("token") else null,
                reason = json.optStringOrNull("reason"),
                fileId = json.optStringOrNull("fileId"),
                name = json.optStringOrNull("name"),
                size = if (json.has("size")) json.optLong("size") else null,
                mime = json.optStringOrNull("mime"),
                sha256 = json.optStringOrNull("sha256"),
                data = json.optStringOrNull("data"),
                ok = if (json.has("ok")) json.optBoolean("ok") else null,
                path = json.optStringOrNull("path"),
            )
        }

        private fun JSONObject.optStringOrNull(key: String): String? =
            if (has(key) && !isNull(key)) optString(key) else null
    }
}

/** `[4 byte big endian length][1 byte kind][body]` framing. */
object Framing {
    fun frame(kind: Byte, body: ByteArray): ByteArray {
        val length = body.size + 1
        val out = ByteArray(4 + length)
        out[0] = (length ushr 24 and 0xFF).toByte()
        out[1] = (length ushr 16 and 0xFF).toByte()
        out[2] = (length ushr 8 and 0xFF).toByte()
        out[3] = (length and 0xFF).toByte()
        out[4] = kind
        body.copyInto(out, 5)
        return out
    }
}

class ProtocolException(message: String) : Exception(message)
